addons/acodec/mp3.c

```diff
@@ -0,0 +1,114 @@
+/*
+ * Allegro5 MP3 reader.
+ * Requires MiniMP3 from https://github.com/lieff/minimp3
+ * author: Mark Watkin (pmprog) 2019
+ */
+
+
+#include "allegro5/allegro.h"
+#include "allegro5/allegro_acodec.h"
+#include "allegro5/allegro_audio.h"
+#include "allegro5/internal/aintern.h"
+#include "allegro5/internal/aintern_audio.h"
+#include "allegro5/internal/aintern_exitfunc.h"
+#include "allegro5/internal/aintern_system.h"
+#include "acodec.h"
+#include "helper.h"
+
+#define MINIMP3_IMPLEMENTATION
+#include <minimp3.h>
+#include <minimp3_ex.h>
+
+ALLEGRO_DEBUG_CHANNEL("acodec")
+
+static bool mp3_libinit = false;
+static mp3dec_t mp3d;
+
+void mp3_initminimp3();
+
+ALLEGRO_SAMPLE *_al_load_mp3(const char *filename)
+{
+   ALLEGRO_FILE *f;
+   ALLEGRO_SAMPLE *spl;
+   ASSERT(filename);
+
+   f = al_fopen(filename, "rb");
+   if (!f)
+      return NULL;
+
+   spl = _al_load_mp3_f(f);
+
+   al_fclose(f);
+
+   return spl;
+}
+
+ALLEGRO_SAMPLE *_al_load_mp3_f(ALLEGRO_FILE *f)
+{
+	mp3_initminimp3(); // Make sure library is initialised
+	
+	mp3dec_file_info_t info;
+	ALLEGRO_SAMPLE *spl = NULL;
+
+	// Read our file size
+	int64_t filesize = al_fsize(f);
+	if (filesize == -1)
+		return NULL;
+
+	// Allocate buffer and read all the file
+	size_t fsize = (size_t)al_fsize(f);
+	uint8_t* mp3data = (uint8_t*)al_malloc(fsize);
+	size_t readbytes = al_fread(f, mp3data, fsize);
+	if (readbytes != fsize)
+	{
+		al_free(mp3data);
+		return NULL;
+	}
+
+	// Decode the file contents, and copy to a new buffer
+	mp3dec_load_buf(&mp3d, mp3data, filesize, &info, NULL, NULL);
+	uint8_t* pcmdata = (uint8_t*)al_malloc(info.samples * sizeof(int16_t));
+	memcpy(pcmdata, info.buffer, info.samples * sizeof(int16_t));
```

Comment:
Why do we need to do a memcpy here. Can we not pass `info.buffer` directly to `al_create_samples`?

Reply:
I think I did this for a reason... but for the life of me, I can't remember why now, so I'll change it and try it out

Reply:
I'm assuming that `info.buffer` would be managed by the `mp3dec` API.  Will Allegro make a copy of the data itself?

Reply:
It's not, because a few lines down you could see that I manually freed info.buffer after copying. Maybe that's what i thought when i originally wrote the code though. Kind of wish I'd made more notes :-)